services/asymkey/deploy_key.go

```diff
@@ -4,26 +4,28 @@
 package asymkey
 
 import (
+	"context"
+
 	"code.gitea.io/gitea/models"
 	asymkey_model "code.gitea.io/gitea/models/asymkey"
 	"code.gitea.io/gitea/models/db"
 	user_model "code.gitea.io/gitea/models/user"
 )
 
 // DeleteDeployKey deletes deploy key from its repository authorized_keys file if needed.
-func DeleteDeployKey(doer *user_model.User, id int64) error {
-	ctx, committer, err := db.TxContext(db.DefaultContext)
+func DeleteDeployKey(ctx context.Context, doer *user_model.User, id int64) error {
+	dbCtx, committer, err := db.TxContext(ctx)
```

Comment:
Not blocking, but we may need a better mechanism for this. If `ctx` is already a transaction, `TxContext` will create a `halfCommitter` and the `Commit` on that does nothing. Only the commit of the outer transaction commits the inner queries. But the original code depends on a finished transaction because the following actions are not reversible.